live-debugger-ffi/src/sender.rs

```diff
@@ -0,0 +1,152 @@
+use std::thread::JoinHandle;
+use tokio::sync::mpsc;
+
+macro_rules! try_c {
+    ($failable:expr) => {
+        match $failable {
+            Ok(o) => o,
+            Err(e) => return MaybeError::Some(ddcommon_ffi::Error::from(format!("{:?}", e))),
+        }
+    };
+}
+
+#[repr(C)]
+pub struct OwnedCharSlice {
+    slice: CharSlice<'static>,
+    free: extern "C" fn(CharSlice<'static>),
+}
+
+unsafe impl Send for OwnedCharSlice {}
+
+impl Drop for OwnedCharSlice {
+    fn drop(&mut self) {
+        (self.free)(self.slice)
+    }
+}
+
+enum SendData {
+    Raw(Vec<u8>),
+    Wrapped(OwnedCharSlice),
+}
+
+async fn sender_routine(
+    endpoint: Arc<Endpoint>,
+    tags: String,
+    mut receiver: mpsc::Receiver<SendData>,
+) {
+    let tags = Arc::new(tags);
+    loop {
+        let data = match receiver.recv().await {
+            None => break,
+            Some(data) => data,
+        };
+
+        let endpoint = endpoint.clone();
+        let tags = tags.clone();
+        tokio::spawn(async move {
+            let data = match &data {
+                SendData::Raw(vec) => vec.as_slice(),
+                SendData::Wrapped(wrapped) => wrapped.slice.as_bytes(),
+            };
+
+            if let Err(e) = sender::send(data, &endpoint, &tags).await {
+                warn!("Failed to send debugger data: {e:?}");
+            }
+        });
+    }
+}
+
+pub struct SenderHandle {
+    join: JoinHandle<()>,
+    channel: mpsc::Sender<SendData>,
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_live_debugger_build_tags(
+    debugger_version: CharSlice,
+    env: CharSlice,
+    version: CharSlice,
+    runtime_id: CharSlice,
+    global_tags: ddcommon_ffi::Vec<Tag>,
+) -> Box<String> {
+    Box::new(generate_tags(
+        &debugger_version.to_utf8_lossy(),
+        &env.to_utf8_lossy(),
+        &version.to_utf8_lossy(),
+        &runtime_id.to_utf8_lossy(),
+        &mut global_tags.into_iter(),
+    ))
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_live_debugger_tags_from_raw(tags: CharSlice) -> Box<String> {
+    Box::new(percent_encode(tags.as_bytes(), CONTROLS).to_string())
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_live_debugger_spawn_sender(
+    endpoint: &Endpoint,
+    tags: Box<String>,
+    handle: &mut *mut SenderHandle,
+) -> MaybeError {
+    let runtime = try_c!(tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build());
+
+    let (tx, mailbox) = mpsc::channel(5000);
+    let endpoint = Arc::new(endpoint.clone());
+
+    *handle = Box::into_raw(Box::new(SenderHandle {
+        join: std::thread::spawn(move || {
+            runtime.block_on(sender_routine(endpoint, *tags, mailbox));
+            runtime.shutdown_background();
+        }),
+        channel: tx,
+    }));
+
+    MaybeError::None
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_live_debugger_send_raw_data(
+    handle: &mut SenderHandle,
+    data: OwnedCharSlice,
+) -> bool {
+    handle.channel.try_send(SendData::Wrapped(data)).is_err()
```

Comment:
I tried using this function and it returns failure for me. The actual error is discarded on this line? How does one troubleshoot why this function is failing?

Reply:
It returns true when there's an error. And false on success. I think that's the issue, i.e. missing an `!`.

But there's just one actual error behind this - when the channel is full, i.e. you're submitting log lines quicker than they can be consumed.